lib/concurrent/configuration.rb

```diff
@@ -1,124 +1,296 @@
 require 'thread'
-require 'concurrent/delay'
+require 'concurrent/lazy_reference'
+require 'concurrent/atomics'
 require 'concurrent/errors'
-require 'concurrent/atomic'
-require 'concurrent/executor/immediate_executor'
-require 'concurrent/executor/thread_pool_executor'
-require 'concurrent/executor/timer_set'
+require 'concurrent/executors'
 require 'concurrent/utility/processor_count'
 
 module Concurrent
   extend Logging
 
-  # A gem-level configuration object.
-  class Configuration
+  # Suppresses all output when used for logging.
+  NULL_LOGGER = lambda { |level, progname, message = nil, &block| }
 
-    # a proc defining how to log messages, its interface has to be:
-    #   lambda { |level, progname, message = nil, &block| _ }
-    attr_accessor :logger
+  # initialize the global executors
+  class << self
+
+    # @!visibility private
+    @@global_logger = Atomic.new(NULL_LOGGER)
+
+    # @!visibility private
+    @@auto_terminate_global_executors = AtomicBoolean.new(true)
+
+    # @!visibility private
+    @@auto_terminate_all_executors = AtomicBoolean.new(true)
+
+    # @!visibility private
+    @@global_fast_executor = LazyReference.new do
+      Concurrent.new_fast_executor(
+        stop_on_exit: @@auto_terminate_global_executors.value)
+    end
+
+    # @!visibility private
+    @@global_io_executor = LazyReference.new do
+      Concurrent.new_io_executor(
+        stop_on_exit: @@auto_terminate_global_executors.value)
+    end
+
+    # @!visibility private
+    @@global_timer_set = LazyReference.new do
+      TimerSet.new(stop_on_exit: @@auto_terminate_global_executors.value)
+    end
+  end
+
+  def self.global_logger
+    @@global_logger.value
```

Comment:
If there is a significant performance difference we can look at constants, but this approach hides direct access to the wrapper object itself. With this approach a user must go through the accessor methods we define, which hides the implementation details, which in turn gives us a greater ability to refactor in the future. The user should not need to know that a global thread pool is wrapped inside of a `LazyReference` object and also know the API of that object. The user should simply be able to post to the global thread pool.